### PR TITLE
Add along-step and tracking cut optical executors

### PR DESCRIPTION
**Pull request by sethrj**
This adds along-step (distance-to-boundary + move + time update) and tracking cut (currently not used, but kills the tracks) to the optical tracking loop.

I fixed an overzealous assertion in the Cherenkov generator: it's allowable not to have any photons generated over a step.

### CODE REVIEW

File: src/celeritas/optical/action/TrackingCutAction.cc [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+TrackingCutAction::TrackingCutAction(ActionId aid)
+    : StaticConcreteAction(
+          aid, "tracking-cut", "kill a track and deposit its energy")
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the boundary action on host.
```
Comment (amandalund) [inline src/celeritas/optical/action/TrackingCutAction.cc:35]:
```suggestion
 * Launch the tracking cut action on host.
```

File: src/celeritas/optical/action/TrackingCutAction.cu [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#include "detail/TrackingCutExecutor.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the boundary action on device.
```
Comment (amandalund) [inline src/celeritas/optical/action/TrackingCutAction.cu]:
```suggestion
 * Launch the tracking cut action on device.
```

File: src/celeritas/optical/action/detail/PropagateExecutor.hh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    // Propagate up to the physics distance
+    real_type step = sim.step_length();
+    CELER_ASSERT(step > 0);
+
+    auto&& geo = track.geometry();
+    Propagation p = geo.find_next_step(step);
+    if (p.boundary)
+    {
+        geo.move_to_boundary();
+        step = p.distance;
```
Comment (amandalund) [inline src/celeritas/optical/action/detail/PropagateExecutor.hh]:
```suggestion
```

**Review by amandalund — APPROVED** [review]
Awesome, thanks for adding these!

**Comment by sethrj** [comment]
Sorry it took so long! I've gotta fix the cuda build and will test it too.

**Pull request merged**